### ISSUE TITLE
[strict-boolean-expressions] Add an `allowDoubleNegation` option

### ISSUE THREAD

**Issue by cherryblossom000** [typescript-eslint/typescript-eslint#2129]
**Repro**
```json
{
  "plugins": ["@typescript-eslint"],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "project": "./tsconfig.json"
  },
  "rules": {
    "@typescript-eslint/strict-boolean-expressions": 2
  }
}
```

`tsconfig.json`
```json
{
  "include": ["**/*.ts"],
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

```TS
declare const nullableBoolean: boolean | undefined
!!nullableBoolean
```

**Expected Result**
No error, because `!!` is being used to explicitly convert `nullableBoolean` into a boolean.

**Actual Result**
```
/path/to/file.ts
  2:3  error  Unexpected nullable boolean value in conditional. Please handle the nullish case explicitly  @typescript-eslint/strict-boolean-expressions

✖ 1 problem (1 error, 0 warnings)
```

**Additional Info**
I propose an `allowDoubleNegation` option to allow unsafe boolean expressions (nullable booleans, nullable strings, nullable numbers, and `any`) when used with `!!`. This could also be configured to only allow some of these expressions such as `allowDoubleNegation: ['nullableBoolean', 'nullableNumber']`.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.0.2` |
| `@typescript-eslint/parser`        | `3.0.2` |
| `TypeScript`                       | `3.9.3` |
| `ESLint`                           | `7.1.0` |
| `node`                             | `12.17.0` |
| `npm`                              | `6.14.5` |

**Comment by bradzacher** [comment]
I don't think this is a good idea for this rule; in the context of what this rule seeks to enforce, this is no stricter or clearer than a single `!`.

The point of the rule is to explicitly and clearly handle every single case. With a double negation (same as with a single negation), you can easily forget to correctly handle the nullish case, and accidentally leak bugs into your code.

**Comment by cherryblossom000** [comment]
@bradzacher That makes sense. I didn't really consider how it was the same as `!`.

However, the reason why I suggested this was because I frequently use optional chaining in boolean contexts like this:
```ts
declare const foo: {bar?: string[]}
if (foo.bar?.includes('baz')) {
  // ...
}
```

Would an `allowOptionalChaining` option be more appropriate or should I just use `foo.bar?.includes('baz') ?? false` to be more explicit?

**Comment by bradzacher** [comment]
You should be explicit to show what you expect the cases to be.

- If you want to treat the nullish case the same as the `false` case, then you can use `nullishBool === true`.
- If you want to treat the nullish case the same as the `true` case, then you can use `nullishBool === false`.

If you do not care about sketchy checks on `boolean`s, you can use the `allowNullableBoolean` option.
